File: 1110/solution.py
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def delNodes(
        self, root: Optional[TreeNode], to_delete: list[int]
    ) -> list[TreeNode]:
        nodes = []
        bad_nodes = set(to_delete)

        def traverse(node, hasParent):
            if not node:
                return None
            if not hasParent and node.val not in bad_nodes:
                nodes.append(node)
            if node.val in bad_nodes:
                traverse(node.left, False)
                traverse(node.right, False)
                return None
            else:
                node.left = traverse(node.left, True)
                node.right = traverse(node.right, True)
            return node

        traverse(root, False)
        return nodes

File: 1110/solution.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<Integer> badNodes = new HashSet<>();
        for (int val : to_delete) {
            badNodes.add(val);
        }
        List<TreeNode> res = new ArrayList<>();
        traverse(root, false, badNodes, res);
        return res;
    }

    private TreeNode traverse(TreeNode node, boolean hasParent, Set<Integer> badNodes, List<TreeNode> res) {
        if (node == null) {
            return null;
        }

        if (!badNodes.contains(node.val)) {
            if (!hasParent) {
                res.add(node);
            }
            node.left = traverse(node.left, true, badNodes, res);
            node.right = traverse(node.right, true, badNodes, res);
        } else {
            traverse(node.left, false, badNodes, res);
            traverse(node.right, false, badNodes, res);
            return null;
        }
        return node;
    }
}

// time O(m + n)
// space O(m + h)
